### PR TITLE
Half-Precision Intrinsics

### PR DESCRIPTION
**Pull request by iyaja**
This adds support for some half-precision intrinsics as mentioned in #391.

The implementations so far:

- Alternative definitions for some functions like `expm1` which did not have a fallback implementation for `Float16` arguments in `Base`. Some `Base` math functions also rely on `libm`, which does not work for the CUDA definitions.
- `shfl_recurse` for `Float16` via `reinterpret`, as in done for `Float64`.
- `atomic_add!` for `Float16`. PTX 6.3, which is the current supported version, only has an `f16` atomic add instruction. Other atomics __can__ be implemented in via `atomic_cas!`, but it seems like like CUDA headers (i.e. `cuda_fp16.h`) only provides a definition for `__half atomicAdd(__half *address, __half val)` and none of the other atomics.

The `device/intrinsics` test suite has also been updated to include the above definitions.

### CODE REVIEW

File: src/device/intrinsics/math.jl [hunk 1]
```diff
@@ -240,9 +249,12 @@ using SpecialFunctions
 # NOTE: CUDA follows fmod, which behaves differently than Base.mod for negative numbers
 #@device_override Base.mod(x::Float64, y::Float64) = ccall("extern __nv_fmod", llvmcall, Cdouble, (Cdouble, Cdouble), x, y)
 #@device_override Base.mod(x::Float32, y::Float32) = ccall("extern __nv_fmodf", llvmcall, Cfloat, (Cfloat, Cfloat), x, y)
+@device_override Base.mod(x::Float16, y::Float16) = Float16(mod(Float32(x), Float32(y)))
+
 
 @device_override Base.rem(x::Float64, y::Float64) = ccall("extern __nv_remainder", llvmcall, Cdouble, (Cdouble, Cdouble), x, y)
 @device_override Base.rem(x::Float32, y::Float32) = ccall("extern __nv_remainderf", llvmcall, Cfloat, (Cfloat, Cfloat), x, y)
+@device_override Base.rem(x::Float16, y::Float16) = Float16(rem(Float32(x), Float32(y)))
```
Comment (maleadt) [inline src/device/intrinsics/math.jl:255]: Why is this required? LLVM seems to handle our `rem_float` intrinsic just fine:

```
julia> CUDA.code_llvm(rem, Tuple{Float16, Float16})
;  @ float.jl:345 within `rem'
define half @julia_rem_1728(half %0, half %1) {
top:
  %2 = frem half %0, %1
  ret half %2
}

julia> CUDA.code_ptx(rem, Tuple{Float16, Float16})
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

        // .globl       julia_rem_1742          // -- Begin function julia_rem_1742
.visible .func  (.param .b32 func_retval0) julia_rem_1742
(
        .param .b32 julia_rem_1742_param_0,
        .param .b32 julia_rem_1742_param_1
)
;
.weak .global .align 8 .u64 julia_rem_1742_slot = julia_rem_1742;
                                        // @julia_rem_1742
.visible .func  (.param .b32 func_retval0) julia_rem_1742(
        .param .b32 julia_rem_1742_param_0,
        .param .b32 julia_rem_1742_param_1
)
{
        .reg .b16       %h<4>;
        .reg .f32       %f<7>;

// %bb.0:                               // %top
        ld.param.b16    %h1, [julia_rem_1742_param_0];
        ld.param.b16    %h2, [julia_rem_1742_param_1];
        cvt.f32.f16     %f1, %h2;
        cvt.f32.f16     %f2, %h1;
        div.rn.f32      %f3, %f2, %f1;
        cvt.rmi.f32.f32         %f4, %f3;
        mul.f32         %f5, %f4, %f1;
        sub.f32         %f6, %f2, %f5;
        cvt.rn.f16.f32  %h3, %f6;
        st.param.b16    [func_retval0+0], %h3;
        ret;
                                        // -- End function
}
```

File: src/device/intrinsics/math.jl [hunk 2]
```diff
@@ -240,9 +249,12 @@ using SpecialFunctions
 # NOTE: CUDA follows fmod, which behaves differently than Base.mod for negative numbers
 #@device_override Base.mod(x::Float64, y::Float64) = ccall("extern __nv_fmod", llvmcall, Cdouble, (Cdouble, Cdouble), x, y)
 #@device_override Base.mod(x::Float32, y::Float32) = ccall("extern __nv_fmodf", llvmcall, Cfloat, (Cfloat, Cfloat), x, y)
+@device_override Base.mod(x::Float16, y::Float16) = Float16(mod(Float32(x), Float32(y)))
```
Comment (maleadt) [inline src/device/intrinsics/math.jl]: Why prefer this above the Base implementation?

File: src/device/intrinsics/math.jl [hunk 3]
```diff
@@ -298,6 +310,7 @@ using SpecialFunctions
 
 @device_override Base.fma(x::Float64, y::Float64, z::Float64) = ccall("extern __nv_fma", llvmcall, Cdouble, (Cdouble, Cdouble, Cdouble), x, y, z)
 @device_override Base.fma(x::Float32, y::Float32, z::Float32) = ccall("extern __nv_fmaf", llvmcall, Cfloat, (Cfloat, Cfloat, Cfloat), x, y, z)
+# @device_override Base.fma(x::Float16, y::Float16, z::Float16) = @asmcall("fma.rnd.f16 \$0, \$1, \$2, \$3", "=h,h,h,h", Float16, Tuple{Float16, Float16, Float16}, x, y, z)
```
Comment (maleadt) [inline src/device/intrinsics/math.jl]: Why disabled?

Comment (iyaja) [reply src/device/intrinsics/math.jl]: I've re-enabled it for now, and it seems to work fine. But it looks like there's no corresponding test case in `device/intrinsics`

File: src/device/intrinsics/math.jl (hunk 2 above)
Comment (iyaja) [reply src/device/intrinsics/math.jl]: You're right - the `Base` version works here. I just removed this.

File: src/device/intrinsics/math.jl (hunk 1 above)
Comment (iyaja) [reply src/device/intrinsics/math.jl:255]: Are you sure? The test case still fails for me without that definition:

```
Error in testset device/intrinsics:
Test Failed at CUDA.jl/test/device/intrinsics.jl:81
  Expression: testf((a->begin
            op.(a, T(2))
        end), T[-1])
ERROR: LoadError: Test run finished with errors
in expression starting at CUDA.jl/test/runtests.jl:490
```

`Base.mod` works though, so I've removed the `CUDA.mod` redefinition (see below).

Comment (maleadt) [reply src/device/intrinsics/math.jl:255]: Ah interesting, it looks like `frem float` behaves differently than `frem half` does...

File: src/device/intrinsics/math.jl (hunk 3 above)
Comment (maleadt) [reply src/device/intrinsics/math.jl]: May be a good time to add one :-)

File: src/device/intrinsics/math.jl [hunk 4]
```diff
@@ -310,7 +308,7 @@ Base.expm1(x::Float16) = Float16(CUDA.expm1(Float32(x)))
 
 @device_override Base.fma(x::Float64, y::Float64, z::Float64) = ccall("extern __nv_fma", llvmcall, Cdouble, (Cdouble, Cdouble, Cdouble), x, y, z)
 @device_override Base.fma(x::Float32, y::Float32, z::Float32) = ccall("extern __nv_fmaf", llvmcall, Cfloat, (Cfloat, Cfloat, Cfloat), x, y, z)
-# @device_override Base.fma(x::Float16, y::Float16, z::Float16) = @asmcall("fma.rnd.f16 \$0, \$1, \$2, \$3", "=h,h,h,h", Float16, Tuple{Float16, Float16, Float16}, x, y, z)
+@device_override Base.fma(x::Float16, y::Float16, z::Float16) = @asmcall("fma.rn.f16 \$0, \$1, \$2, \$3;", "=h,h,h,h", true, Float16, Tuple{Float16, Float16, Float16}, x, y, z)
```
Comment (vchuravy) [inline src/device/intrinsics/math.jl]: Is there no LLVM intrinsics for this? Does ` @llvm.fma.f16` expand to the right thing?

Comment (iyaja) [reply src/device/intrinsics/math.jl]: Looks like it does, and it passes the new tests I've written for `CUDA.fma`. I'm not entirely sure which one is more performant (or if there's even a difference) but I've changed it for now.

File: src/device/intrinsics/math.jl [hunk 5]
```diff
@@ -308,7 +308,17 @@ Base.expm1(x::Float16) = Float16(CUDA.expm1(Float32(x)))
...
 @device_override Base.fma(x::Float64, y::Float64, z::Float64) = ccall("extern __nv_fma", llvmcall, Cdouble, (Cdouble, Cdouble, Cdouble), x, y, z)
 @device_override Base.fma(x::Float32, y::Float32, z::Float32) = ccall("extern __nv_fmaf", llvmcall, Cfloat, (Cfloat, Cfloat, Cfloat), x, y, z)
-@device_override Base.fma(x::Float16, y::Float16, z::Float16) = @asmcall("fma.rn.f16 \$0, \$1, \$2, \$3;", "=h,h,h,h", true, Float16, Tuple{Float16, Float16, Float16}, x, y, z)
+
+@device_override function Base.fma(x::Float16, y::Float16, z::Float16)
+    mod = """
+        declare half @llvm.fma.f16(half %a, half %b, half %c)
+
+        define half @entry(half %0, half %1, half %2) #0 {
+            %rv = call half @llvm.fma.f16(half %0, half %1, half %2)
```
Comment (maleadt) [inline src/device/intrinsics/math.jl]: Since there's nothing happening here other than directly calling the intrinsic, you can `ccall` with the `llvmcall` calling convention.

Comment (iyaja) [reply src/device/intrinsics/math.jl]: Ok, fixed.

File: src/device/intrinsics/math.jl (hunk 3 above)
Comment (iyaja) [reply src/device/intrinsics/math.jl]: Added. See below.

**Comment by maleadt** [comment]
All green :-) Let's merge this.
